src/font.rs

```diff
@@ -102,15 +100,271 @@ struct Loca<'a>(&'a [u8]);
+            let end_value = get_u16(self.0, search as usize).unwrap();
+            if end_value >= code_point {
+                let start_pos = Self::get_start_counts_position(seg_count) + 2 * index;
+                let start_value = get_u16(self.0, start_pos as usize).unwrap();
+                if start_value > code_point {
+                    size /= 2;
+                    index -= size;
+                } else {
+                    return self.extract_glyph_id(code_point, start_value, seg_count, index);
+                }
+            } else {
+                size /= 2;
+                index += size;
+            }
+        }
+        None
+    }
+}
+
+impl<'a> Debug for EncodingFormat4<'a> {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        f.debug_struct("EncodingFormat4")
+            .field("format", &self.get_format())
+            .field("length", &self.get_length())
+            .field("language", &self.get_language())
+            .field("segCountX2", &self.get_seg_count_x_2())
+            .field("searchRange", &self.get_search_range())
+            .field("entrySelector", &self.get_entry_selector())
+            .field("rangeShift", &self.get_range_shift())
+            .field("endCounts", &self.get_end_counts())
+            .field("startCounts", &self.get_start_counts())
+            .field("idDeltas", &self.get_id_deltas())
+            .field("idRangeOffsets", &self.get_id_range_offsets())
+            .finish()
+    }
+}
+
+struct Cmap<'a>(&'a [u8]);
+
+impl<'a> Cmap<'a> {
+    fn get_version(&'a self) -> u16 {
+        get_u16(self.0, 0).unwrap()
+    }
+
+    fn get_num_tables(&'a self) -> u16 {
+        get_u16(self.0, 2).unwrap()
+    }
+
+    fn get_encoding_record(&'a self, index: u16) -> Option<EncodingRecord<'a>> {
+        if index >= self.get_num_tables() {
+            return None;
+        }
+        let enc_offset = (index * 8 + 4) as usize;
+        let encoding_data = &self.0[enc_offset as usize..(enc_offset + 12) as usize];
+        Some(EncodingRecord(encoding_data))
+    }
+
+    fn get_encoding_records(&'a self) -> Vec<EncodingRecord> {
+        let mut encodings = vec![];
+        for i in 0..self.get_num_tables() {
+            encodings.push(self.get_encoding_record(i).unwrap());
+        }
+        encodings
+    }
+
+    fn get_encoding(&'a self, index: u16) -> Option<EncodingFormat4<'a>> {
+        if index >= self.get_num_tables() {
+            return None;
+        }
+        let record = self.get_encoding_record(index).unwrap();
+        let subtable_len = get_u16(self.0, (record.get_offset() + 2) as usize).unwrap() as u32;
+        let encoding_data =
+            &self.0[record.get_offset() as usize..(record.get_offset() + subtable_len) as usize];
+        Some(EncodingFormat4(encoding_data))
+    }
+
+    fn get_encodings(&'a self) -> Vec<EncodingFormat4> {
+        let mut encodings = vec![];
+        for i in 0..self.get_num_tables() {
```

Comment:
So this seems a little wrong to me because it looks like it will create `EncodingFormat4` structs for encoding tables that aren't format 4, which in turn will cause confusing output and possibly panics when debug printing them. My preferred approach would be for the format to be runtime-checked at the point where the more specific type is created (ie in `get_encoding`), then the code here tweaked to handle the `None` case.